File: PHP/Gun/Gun.php
<?php
    function myAutoload($className) {
        $path = "Exceptions/" . $className . ".php";
        include($path);
    }
    spl_autoload_register("myAutoload");
    
    class Gun {
        const SHOT = 1;
        const BULLET = 1;
        
        private $model;
        private $totalShots;
        private $isReady;
        private $amount, $capacity;
        
        private function isString($value) {
            if ( is_string($value) ) {
                return $value;
            }
            throw new InvalidParam("Invalid param '$value'.");
        }
        
        private function isNumber($value) {
            if ( is_int($value) ) {
                return $value;
            }
            throw new InvalidParam("Invalid param '$value'.");
        }
        
        public function __construct($model = "Beretta", $capacity = 8) {
            $this->model = $this->isString($model);
            $this->capacity = $this->isNumber($capacity);
            $this->amount = 0;
            $this->totalShots = 0;
            $this->isReady = false; 
        }
        
        public function __get($name) {
            if ( property_exists($this, $name) ) {
                return $this->$name;
            }
            throw new InvalidAttribute("Attribute error: attribute '$name' not found.");
        }
        
        public function prepare() {
            $this->isReady = !$this->isReady;
        }
        
        public function reload() {
            if ( $this->capacity <= 0 ) {
                echo "You entered incorrect magazine capacity!";
                return;
            }
            $this->amount = $this->capacity;
        }
        
        public function shoot() {
            if ( !$this->isReady ) {
                throw new NotReadyToShoot("Please, prepare your gun for shooting!");
            }
            if ( $this->amount == 0 ) {
                throw new OutOfRounds("Please, reload your gun to continue shooting!");
            }
            echo "Bang!";
            $this->amount -= self::BULLET;
            $this->totalShots += self::SHOT;
        }
        
        public function __toString() {
            $convertedIsReady = $this->isReady ? 'true' : 'false';
            
            return "Characteristics of the gun:" . PHP_EOL 
                . "The gun model - " . $this->model . "." . PHP_EOL
                . "Ready to shoot - " . $convertedIsReady . "." . PHP_EOL
                . "Bullets quantity - " . $this->amount . "." . PHP_EOL
                . "Magazine capacity - " . $this->capacity . "." . PHP_EOL
                . "Total shots - " . $this->totalShots . ".";
        }
    }
?>

File: W5/Gun/Gun.py
from Exceptions import OutOfRounds, NotReadyError, GunCapacityError


class Gun:
    _BULLET = 1
    _SHOOT = 1

    def __init__(self, model: str = 'Beretta', capacity: int = 8) -> None:
        self._amount = 0

        if capacity <= 0:
            raise GunCapacityError('You entered incorrect magazine capacity!')
        self._capacity = int(capacity)

        self._is_ready = False
        self._model = str(model)
        self._total_shots = 0

    def __str__(self) -> str:
        return f'{self._model}: {self._is_ready}\n' \
               f'{self._total_shots}: {self._capacity}/{self._amount}'

    def __repr__(self) -> str:
        return f'{self._model}: {self._is_ready}\n' \
               f'{self._total_shots}: {self._capacity}/{self._amount}'

    @property
    def amount(self) -> int:
        return self._amount

    @property
    def capacity(self) -> int:
        return self._capacity

    @property
    def is_ready(self) -> bool:
        return self._is_ready

    @property
    def model(self) -> str:
        return self._model

    @property
    def total_shots(self) -> int:
        return self._total_shots

    def prepare(self) -> None:
        self._is_ready = not self._is_ready

    def reload(self) -> None:
        self._amount = self._capacity

    def shoot(self) -> None:
        if not self._is_ready:
            raise NotReadyError('Please, prepare for shooting!')
        if self._amount == 0:
            raise OutOfRounds('Please, reload your gun to continue shooting!')

        print('Bang!')
        self._amount -= Gun._BULLET
        self._total_shots += Gun._SHOOT
